proposals/0047-rerouting.md

```diff
@@ -0,0 +1,186 @@
+<!--
+  Note: You are probably looking for `stage-1--discussion-template.md`!
+  This template is reserved for anyone championing an already-approved proposal.
+
+  Community members who would like to propose an idea or feature should begin
+  by creating a GitHub Discussion. See the repo README.md for more info.
+
+  To use this template: create a new, empty file in the repo under `proposals/${ID}.md`.
+  Replace `${ID}` with the official accepted proposal ID, found in the GitHub Issue
+  of the accepted proposal.
+-->
+
+- Start Date: 2024-04-17
+- Reference Issues: https://github.com/withastro/roadmap/issues/896
+- Implementation PR: 
+
+# Summary
+
+Programmatic control over routing.
+
+# Background & Motivation
+
+
+As astro grows in popularity, it is being used for more and more use cases. One of these is multitenant websites where a tenant may be given a subdomain. This proposal introduces a powerful primitive that allows astro users to build websites with complex routing requirements. Rather than a limitation, file-based routing is now a sensible default!
```

Comment:
I don't think I follow this explanation of motivation. What does subdomains have to do with this feature?

Reroute here is like a rewrite rule that has been popular with tradition HTTP servers for a long time, for example with nginx: https://www.nginx.com/blog/creating-nginx-rewrite-rules/

One common motivation is to support routes being accessible from multiple URLs. That's why you might use this feature in middleware.

Another use-case specific to Astro is wanting to render a different page with the same URL, for example to show a login page when the user is missing the login cookie.